src/material-experimental/mdc-form-field/form-field.ts

```diff
@@ -0,0 +1,462 @@
+    // controls. MDC needs a reference to the native input optionally to handle character
+    // counting and value updating. These are both things we do not handle from within the
+    // form-field, so we can just return null.
+    getNativeInput: () => null,
+
+    // This method will never be called since we do not have the ability to add event listeners
+    // to the native input. This is because the form control is not necessarily an input, and
+    // the form field deals with abstract form controls of any type.
+    setLineRippleTransformOrigin: () => {},
+
+    // The foundation tries to register click and keyboard events on the form-field to figure out
+    // if the input value changes through user interaction. Based on that, the foundation tries
+    // to focus the input. Since we do not handle the input value as part of the form-field, nor
+    // it's guaranteed to be an input (see adapter methods above), this is a noop.
+    deregisterTextFieldInteractionHandler: () => {},
+    registerTextFieldInteractionHandler: () => {},
+
+    // The foundation tries to setup a "MutationObserver" in order to watch for attributes
+    // like "maxlength" or "pattern" to change. The foundation will update the validity state
+    // based on that. We do not need this logic since we handle the validity through the
+    // abstract form control instance.
+    deregisterValidationAttributeChangeHandler: () => {},
+    registerValidationAttributeChangeHandler: () => null as any,
+  };
+
+  constructor(private _elementRef: ElementRef,
+              private _changeDetectorRef: ChangeDetectorRef,
+              @Optional() @Inject(MAT_FORM_FIELD_DEFAULT_OPTIONS)
+              private _defaults?: MatFormFieldDefaultOptions,
+              @Optional() @Inject(MAT_LABEL_GLOBAL_OPTIONS) private _labelOptions?: LabelOptions,
+              @Optional() @Inject(ANIMATION_MODULE_TYPE) _animationMode?: string) {
+    this._animationsEnabled = _animationMode !== 'NoopAnimations';
+
+    if (_defaults && _defaults.appearance) {
+      this.appearance = _defaults.appearance;
+    } else if (_defaults && _defaults.hideRequiredMarker) {
+      this.hideRequiredMarker = true;
+    }
+  }
+
+  ngAfterViewInit() {
+    this._foundation = new MDCTextFieldFoundation(this._adapter);
+
+    Object.defineProperty(this._foundation, 'shouldFloat', {
+      get: () => this._shouldLabelFloat(),
+    });
+
+    // Initial focus state sync. This happens rarely, but we want to account for
+    // it in case the form-field control has "focused" set to true on init.
+    this._updateFocusState();
+    // Initial notch update since we overwrote the "shouldFloat" getter.
+    this._rerenderOutlineNotch();
+    // Enable animations now. This ensures we don't animate on initial render.
+    this._subscriptAnimationState = 'enter';
+  }
+
+  ngAfterContentInit() {
+    this._assertFormFieldControl();
+    const control = this._control;
```

Comment:
Could you move the bulk of the logic in this method into one or more separate methods that get called from here? In general, I want to push more logic out of lifecycle hooks directly and into more descriptively named methods

Reply:
yup. done